### PR TITLE
Upgrade etcd dependency to v3.3.15

### LINKED ISSUE #4349: Vault with etcd backend storage with multiple adresses uses only the first server's ServerName for all endpoints, generating bad cert tls error on etcd side
**Environment:**

* Vault Version: Vault v0.10.0 ('5dd7f25f5c4b541f2da62d70075b6f82771a650d')
* Etcd Version: v3.3.3
* Operating System/Architecture: Ubuntu 16.04/x86

**Vault Config File:**

```hcl
storage "etcd" {
  address = "https://etcd-0.example.com:2379,https://etcd-1.example.com:2379,https://etcd-2.example.com:2379"
  etcd_api = "v3"
  ha_enabled = "true"
  path = "vault/"
  tls_ca_file = "/etcd/rootCA.crt"
  tls_cert_file = "/etcd/client.crt"
  tls_key_file = "/etcd/client.key"  
}

listener "tcp" {
 address     = "127.0.0.1:8200"
 tls_disable = "true"
}

api_addr = "http://127.0.0.1:8200"
cluster_addr = "http://127.0.0.1:8201"
```

**Startup Log Output:**

```
==> Vault server configuration:

             Api Address: http://127.0.0.1:8200
                     Cgo: disabled
         Cluster Address: https://127.0.0.1:8201
              Listener 1: tcp (addr: "127.0.0.1:8200", cluster address: "127.0.0.1:8201", tls: "disabled")
               Log Level: debug
                   Mlock: supported: true, enabled: true
                 Storage: etcd (HA available)
                 Version: Vault v0.10.0
             Version Sha: 5dd7f25f5c4b541f2da62d70075b6f82771a650d

==> Vault server started! Log data will stream in below:

2018-04-13T16:15:52.200+0700 [DEBUG] storage.cache: creating LRU cache: size=0
2018-04-13T16:15:52.208+0700 [DEBUG] cluster listener addresses synthesized: cluster_addresses=[127.0.0.1:8201]
```

**Expected Behavior:**

All the etcd endpoints don't throw tls error and accept the connection

**Actual Behavior:**

Two last etcd endpoints thrown tls error, stating bad certificate, thus rejecting them.
I suspect that this is because Vault sends request to them using the first server's (etcd-0) server name.

**Sample etcd error logs: (from etcd-1. etcd-2 showed the same error)**
```
2018-04-13 08:01:50.267438 I | embed: rejected connection from "10.10.145.64:32946" (error "remote error: tls: bad certificate", ServerName "etcd-0.example.com")
2018-04-13 08:05:30.852145 I | embed: rejected connection from "10.10.145.64:33136" (error "remote error: tls: bad certificate", ServerName "etcd-0.example.com")
2018-04-13 08:05:30.932451 I | embed: rejected connection from "10.10.145.64:33144" (error "remote error: tls: bad certificate", ServerName "etcd-0.example.com")
```

**Steps to Reproduce:**

- Create an etcd cluster of three endpoints with client-tls enabled
- Implement vault with etcd backend storage with multiple endpoints
- Start up the server
- See the etcd error

### LINKED ISSUE #4961: Vault is inaccessible if an etcd unit is lost
**Describe the bug**
When vault is deployed in a cluster, using etcd for HA management, losing the first etcd unit in the 'address' list results in vault returning 500s 

**To Reproduce**
- Deploy 3 units of vault. In my test case use 1 mysql unit for storage and 3 units of etcd for ha_storage.
- Initialise and unseal vault and validate API with "vault status"
- Shutdown the etcd unit that corresponds to the first IP in the 'address' list (in the 'ha_storage "etcd" ' section)
- Run "vault status" with VAULT_ADDR pointing at any of the vault units

```$ vault status
Error checking leader status: Error making API request.

URL: GET http://10.53.82.200:8200/v1/sys/leader
Code: 500. Errors:

* context deadline exceeded
```

**Expected behavior**
I would expect vault to continue to function and respond to API requests, without interruption, in the event of losing an etcd unit. 

**Environment:**
* Vault Server Version: 0.10.3
* Vault Client Version: 0.10.1
* Server OS: Ubuntu 16.04.5 (Xenial) x86_64

Vault server configuration file(s):

```hcl
api_addr = "http://10.53.82.113:8200"
cluster_addr = "http://10.53.82.113:8201"
disable_mlock = true
storage "mysql" {
  username = "vault"
  password = "kpzLkcjfmX9w45n542zwLyrJBppfg5rP"
  database = "vault"
  address = "10.53.82.226:3306"
}
ha_storage "etcd" {
  ha_enabled = "true"
  address = "https://10.53.82.119:2379,https://10.53.82.150:2379,https://10.53.82.157:2379"
  tls_ca_file = "/var/snap/vault/common/etcd-ca.pem"
  tls_cert_file = "/var/snap/vault/common/etcd-cert.pem"
  tls_key_file = "/var/snap/vault/common/etcd.key"
  etcd_api = "v3"
}
listener "tcp" {
  address = "0.0.0.0:8200"
  tls_disable = 1
}

# Localhost only listener for charm access to vault.
listener "tcp" {
  address = "127.0.0.1:8220"
  tls_disable = 1
}
```

**Additional context**

On the active vault unit the following entries appear in its log when the etcd unit goes down:

```
2018-07-20T08:24:16.614Z [WARN ] core: leadership lost, stopping active operation
2018-07-20T08:24:16.614Z [INFO ] core: pre-seal teardown starting
2018-07-20T08:24:16.614Z [INFO ] core: stopping cluster listeners
2018-07-20T08:24:16.614Z [INFO ] core: shutting down forwarding rpc listeners
2018-07-20T08:24:16.614Z [INFO ] core: forwarding rpc listeners stopped
2018-07-20T08:24:16.907Z [INFO ] core: rpc listeners successfully shut down
2018-07-20T08:24:16.907Z [INFO ] core: cluster listeners successfully shut down
2018-07-20T08:24:16.907Z [INFO ] rollback: stopping rollback manager
2018-07-20T08:24:16.908Z [INFO ] core: pre-seal teardown complete
```

The standby vault units do not log anything as a result of the lost etcd unit.

### PR DESCRIPTION
**Pull request by jsok**
Contains an important fix in clientv3 that allows vault to
successfully failover to another etcdv3 endpoint in the event that the
current active connection becomes unavailable.

See also:
 * https://github.com/etcd-io/etcd/issues/9949
 * https://github.com/etcd-io/etcd/pull/10911
 * https://github.com/etcd-io/etcd/blob/master/CHANGELOG-3.3.md#v3314-2019-08-16

Fixes #4961
Fixed #4349

### CODE REVIEW

**Comment by jsok** [comment]
Looks like ASF git is having issues right now: https://status.apache.org/incidents/63030p4241xj

**Comment by xtrusia** [comment]
Hello, Have you tested this commit your test env?
I'm interested in this commit as well. and I'm looking forward to this PR can be merged.

**Comment by xtrusia** [comment]
> Hello, Have you tested this commit your test env?
> I'm interested in this commit as well. and I'm looking forward to this PR can be merged.
> 
> ADDED 1
> 
> I still have the issue jsok mentioned.
> 
> fatal: unable to access 'https://git.apache.org/thrift.git/': Failed to connect to git.apache.org port 443: Connection timed out
> 
> ADDED 2
> I solved #1 with export GOPROXY setting but I needed to set below version instead of 0.5.0
> 
> google.golang.org/api v0.6.0

Even If I use this branch, It is not working great after shutting down one etcd node.
it takes over 20mins to recover.

**Comment by jefferai** [comment]
I get a _very_ different set of go.mod changes when I run `go get go.etcd.io/etcd`.

**Comment by jsok** [comment]
> I get a _very_ different set of go.mod changes when I run `go get go.etcd.io/etcd`.

vendoring etcd is a bit confusing to me as well - i've tried to pull `master` in at the commish where 3.3.15 was released, as opposed to using the 3.3.15 tag as between 3.3.14 and 3.3.15 they [flip flopped between glide, go.mod and back to glide again](https://github.com/etcd-io/etcd/commit/aa1e17aac3a6cad4dd1f62578ffe4c872b021693) 😕 

We also don't want to just pull from HEAD of master as that's now 3.4.x stream of code and probably larger an upgrade than we need for this fix.

**Comment by jefferai** [comment]
If you pull in from the 3.3.15 tag it is still marked as incompatible but does update deps. I think that's likely the right path for now, assuming you can verify that the pulled in code works properly.

File: go.mod [hunk 1]
```diff
@@ -42,7 +42,7 @@ require (
 	github.com/go-test/deep v1.0.2
```
Comment (xtrusia) [inline go.mod]: for thrift error, as master branch, we need to put below before require block

replace git.apache.org/thrift.git => github.com/apache/thrift v0.12.0

**Comment by jefferai** [comment]
Note that things don't actually build successfully just by updating the deps. Making them build likely means essentially reverting the work that was done to adapt to the new APIs and standardizing on the github.com/coreos/etcd repo names, and then changing them back later.

The likely better approach is to just go to v3.4.1. Updating to that (and an undeclared but necessary update to google.golang.org/api) gives the following set, which is far more manageable, but also has some deep changes (prometheus, grpc):
```
diff --git a/go.mod b/go.mod
index befb49285..0c091b7a4 100644
--- a/go.mod
+++ b/go.mod
@@ -110,22 +110,22 @@ require (
        github.com/pkg/errors v0.8.1
        github.com/posener/complete v1.2.1
        github.com/pquerna/otp v1.1.0
-       github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
-       github.com/prometheus/common v0.2.0
+       github.com/prometheus/client_golang v1.0.0
+       github.com/prometheus/common v0.4.1
        github.com/ryanuber/columnize v2.1.0+incompatible
        github.com/ryanuber/go-glob v1.0.0
        github.com/samuel/go-zookeeper v0.0.0-20180130194729-c4fab1ac1bec
        github.com/shopspring/decimal v0.0.0-20180709203117-cd690d0c9e24 // indirect
        github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94 // indirect
        github.com/stretchr/testify v1.3.0
-       go.etcd.io/bbolt v1.3.2
-       go.etcd.io/etcd v0.0.0-20190412021913-f29b1ada1971
+       go.etcd.io/bbolt v1.3.3
+       go.etcd.io/etcd v0.0.0-20190917205325-a14579fbfb1a
        golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
        golang.org/x/net v0.0.0-20190813141303-74dc4d7220e7
-       golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
-       google.golang.org/api v0.5.0
+       golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
+       google.golang.org/api v0.10.0
        google.golang.org/genproto v0.0.0-20190801165951-fa694d86fc64
-       google.golang.org/grpc v1.22.0
+       google.golang.org/grpc v1.23.1
        gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce
        gopkg.in/ory-am/dockertest.v3 v3.3.4
        gopkg.in/square/go-jose.v2 v2.3.1
```
@chrishoffman @briankassouf please weigh in

**Comment by xtrusia** [comment]
This is working well in my test.

3 etcd 3.3.15 servers + vault master built with below diff, 

shutting down etcd server one by one and  check 
$etcdctl endpoint status --cluster
$vault secrets list -detailed.

> Note that things don't actually build successfully just by updating the deps. Making them build likely means essentially reverting the work that was done to adapt to the new APIs and standardizing on the github.com/coreos/etcd repo names, and then changing them back later.
> 
> The likely better approach is to just go to v3.4.1. Updating to that (and an undeclared but necessary update to google.golang.org/api) gives the following set, which is far more manageable, but also has some deep changes (prometheus, grpc, bbolt):
> 
> ```
> diff --git a/go.mod b/go.mod
> index befb49285..0c091b7a4 100644
> --- a/go.mod
> +++ b/go.mod
> @@ -110,22 +110,22 @@ require (
>         github.com/pkg/errors v0.8.1
>         github.com/posener/complete v1.2.1
>         github.com/pquerna/otp v1.1.0
> -       github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
> -       github.com/prometheus/common v0.2.0
> +       github.com/prometheus/client_golang v1.0.0
> +       github.com/prometheus/common v0.4.1
>         github.com/ryanuber/columnize v2.1.0+incompatible
>         github.com/ryanuber/go-glob v1.0.0
>         github.com/samuel/go-zookeeper v0.0.0-20180130194729-c4fab1ac1bec
>         github.com/shopspring/decimal v0.0.0-20180709203117-cd690d0c9e24 // indirect
>         github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94 // indirect
>         github.com/stretchr/testify v1.3.0
> -       go.etcd.io/bbolt v1.3.2
> -       go.etcd.io/etcd v0.0.0-20190412021913-f29b1ada1971
> +       go.etcd.io/bbolt v1.3.3
> +       go.etcd.io/etcd v0.0.0-20190917205325-a14579fbfb1a
>         golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
>         golang.org/x/net v0.0.0-20190813141303-74dc4d7220e7
> -       golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
> -       google.golang.org/api v0.5.0
> +       golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
> +       google.golang.org/api v0.10.0
>         google.golang.org/genproto v0.0.0-20190801165951-fa694d86fc64
> -       google.golang.org/grpc v1.22.0
> +       google.golang.org/grpc v1.23.1
>         gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce
>         gopkg.in/ory-am/dockertest.v3 v3.3.4
>         gopkg.in/square/go-jose.v2 v2.3.1
> ```
> 
> @chrishoffman @briankassouf please weigh in

**Comment by xtrusia** [comment]
hmm.. it doesn;t work well on 2nd test.

> This is working well in my test.
> 
> 3 etcd 3.3.15 servers + vault master built with below diff,
> 
> shutting down etcd server one by one and check
> $etcdctl endpoint status --cluster
> $vault secrets list -detailed.
> 
> > Note that things don't actually build successfully just by updating the deps. Making them build likely means essentially reverting the work that was done to adapt to the new APIs and standardizing on the github.com/coreos/etcd repo names, and then changing them back later.
> > The likely better approach is to just go to v3.4.1. Updating to that (and an undeclared but necessary update to google.golang.org/api) gives the following set, which is far more manageable, but also has some deep changes (prometheus, grpc, bbolt):
> > ```
> > diff --git a/go.mod b/go.mod
> > index befb49285..0c091b7a4 100644
> > --- a/go.mod
> > +++ b/go.mod
> > @@ -110,22 +110,22 @@ require (
> >         github.com/pkg/errors v0.8.1
> >         github.com/posener/complete v1.2.1
> >         github.com/pquerna/otp v1.1.0
> > -       github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
> > -       github.com/prometheus/common v0.2.0
> > +       github.com/prometheus/client_golang v1.0.0
> > +       github.com/prometheus/common v0.4.1
> >         github.com/ryanuber/columnize v2.1.0+incompatible
> >         github.com/ryanuber/go-glob v1.0.0
> >         github.com/samuel/go-zookeeper v0.0.0-20180130194729-c4fab1ac1bec
> >         github.com/shopspring/decimal v0.0.0-20180709203117-cd690d0c9e24 // indirect
> >         github.com/streadway/amqp v0.0.0-20190404075320-75d898a42a94 // indirect
> >         github.com/stretchr/testify v1.3.0
> > -       go.etcd.io/bbolt v1.3.2
> > -       go.etcd.io/etcd v0.0.0-20190412021913-f29b1ada1971
> > +       go.etcd.io/bbolt v1.3.3
> > +       go.etcd.io/etcd v0.0.0-20190917205325-a14579fbfb1a
> >         golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
> >         golang.org/x/net v0.0.0-20190813141303-74dc4d7220e7
> > -       golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
> > -       google.golang.org/api v0.5.0
> > +       golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
> > +       google.golang.org/api v0.10.0
> >         google.golang.org/genproto v0.0.0-20190801165951-fa694d86fc64
> > -       google.golang.org/grpc v1.22.0
> > +       google.golang.org/grpc v1.23.1
> >         gopkg.in/mgo.v2 v2.0.0-20180705113604-9856a29383ce
> >         gopkg.in/ory-am/dockertest.v3 v3.3.4
> >         gopkg.in/square/go-jose.v2 v2.3.1
> > ```
> > 
> > 
> > @chrishoffman @briankassouf please weigh in

**Comment by xtrusia** [comment]
test env,
3 etcds, 1 vault
shut down etcd node one by one ( even if it is leader or not ) several times

I built test program which is keep asking member list as below
https://pastebin.ubuntu.com/p/pcynNM4W3W/

built vault with above dependency changes and added 
DialOptions: []grpc.DialOption{grpc.WithBlock()},

my test program works well even if I shutdown one of etcd node one by one several times 
https://pastebin.ubuntu.com/p/HmznhGK5Rx/
there is one context deadline exceeded but it is recovered soon

but vault got error and didn't recovered well ( it is checked with "vault secrets list -detailed" ), it took around 20mins

vault log
https://pastebin.ubuntu.com/p/5RbB6FBwr3/

**Comment by michelvocks** [comment]
Hi @jsok!

We are interested in getting this merged soon. Would you be able to address the review comments mentioned above by @jefferai in the near term? If not, we can also overtake this issue for you.

Please let us know what you think.

Cheers,
Michel

**Comment by jsok** [comment]
> If not, we can also overtake this issue for you

@michelvocks that would be fine, I don't think I'll get to it until early next year at this rate.

Would also be worth jumping to 3.3.18 too potentially.

**Comment by michelvocks** [comment]
Hi @jsok!

Thank you for your quick reply. 
I close this PR in favor of https://github.com/hashicorp/vault/pull/8037.

Cheers,
Michel

**Pull request closed without merging**